src/compiler.jl

```diff
@@ -279,101 +281,206 @@ end
+
+struct Merger{seen,worldT,justActive,UnionSret}
+    world::worldT
+end
+
+@inline element(::Val{T}) where T = T
+
+@inline function (c::Merger{seen,worldT,justActive,UnionSret})(f::Int) where {seen,worldT,justActive,UnionSret}
+    T = element(first(seen))
+
+    if justActive && ismutabletype(T)
+        return Val(AnyState)
+    end
+
+    subT = fieldtype(T, f)
+
+    if justActive && !allocatedinline(subT)
+        return Val(AnyState)
+    end
+
+    sub = active_reg_inner(subT, seen, c.world, Val(justActive), Val(UnionSret))
+
+    if sub == AnyState
+        Val(AnyState)
+    else
+        if sub == DupState
+            if justActive
+                Val(AnyState)
+            else
+                Val(DupState)
+            end
+        else
+            Val(sub)
+        end
     end
-    return DupState
 end
-@inline function active_reg_inner(PT::Type{<:Array{T}}, seen) where {T}
-    state = active_reg_inner(T, seen)
-    if state == AnyState
-        return AnyState
+
+@inline forcefold(::Val{RT}) where RT = RT
+
+@inline function forcefold(::Val{ty}, ::Val{sty}, C::Vararg{Any, N}) where {ty, sty, N}
+    if sty == AnyState || sty == ty
+        return forcefold(Val(ty), C...)
+    end
+    if ty == AnyState
+        return forcefold(Val(sty), C...)
+    else
+        return MixedState
     end
-    return DupState
 end
 
-@inline function active_reg_inner(::Type{T}, seen) where T
-    if T ∈ keys(seen)
-        return seen[T]
+@inline function active_reg_inner(::Type{T}, seen::ST, world::Union{Nothing, UInt}, ::Val{justActive}=Val(false), ::Val{UnionSret}=Val(false))::ActivityState where {ST,T, justActive, UnionSret}
+
+    if T === Any
+        return DupState
     end
-    if EnzymeRules.inactive_type(T)
-        return seen[T] = AnyState
+
+    if T === Union{}
+        return AnyState
+    end
+
+    if T <: Complex
+        return active_reg_inner(ptreltype(T), seen, world, Val(justActive), Val(UnionSret))
     end
+
+    if T <: AbstractFloat
+        return ActiveState
+    end
+
+    if T <: Ptr || T <: Core.LLVMPtr || T <: Base.RefValue || T <: Array || isa(T, Type{Tuple{Vararg{T}}} where T)
+        if justActive
+            return AnyState
+        end
+        if active_reg_inner(ptreltype(T), seen, world, Val(justActive), Val(UnionSret)) == AnyState
+            return AnyState
+        else
+            return DupState
+        end
+    end
+
+    if T <: Integer
+        return AnyState
+    end
+
     if isghostty(T) || Core.Compiler.isconstType(T)
         return AnyState
     end
+
+    inactivety = if typeof(world) === Nothing
+        EnzymeCore.EnzymeRules.inactive_type(T)
+    else
+        inmi = GPUCompiler.methodinstance(typeof(EnzymeCore.EnzymeRules.inactive_type), Tuple{Type{T}}, world)
+        args = Any[EnzymeCore.EnzymeRules.inactive_type, T];
+        ccall(:jl_invoke, Any, (Any, Ptr{Any}, Cuint, Any), EnzymeCore.EnzymeRules.inactive_type, args, length(args), inmi)
```

Comment:
@wsmoses what did you want to do here? Did you want to do an `invoke_in_world`?

Reply:
yeah basically to run inactive_type on the given world (which may almost certainly be later than the given world of creation of that method)